patches/tModLoader/Terraria/ModLoader/PosData.cs

```diff
@@ -0,0 +1,150 @@
+using Microsoft.Xna.Framework;
+using System;
+using System.Collections.Generic;
+
+namespace Terraria.ModLoader
+{
+	public readonly struct PosData<T>
+	{
+		public class SparseLookupBuilder 
+		{
+			private readonly List<PosData<T>> list;
+			private PosData<T> last;
+
+			public SparseLookupBuilder(int capacity = 1048576) {
+				list = new List<PosData<T>>(capacity);
+				last = new PosData<T>(-1, default);
+			}
+
+			public void Add(int pos, T value) {
+				if (pos <= last.pos)
+					throw new ArgumentException($"Must build in ascending index order. Prev: {last.pos}, pos: {pos}");
+
+				if (!EqualityComparer<T>.Default.Equals(value, last.value))
+					list.Add(last = new PosData<T>(pos, value));
+			}
+
+			public PosData<T>[] Build() => list.ToArray();
+		}
+
+		public class SparseLookupReader
+		{
+			private readonly PosData<T>[] data;
+			private PosData<T> current;
+			private int nextIdx;
+
+			public SparseLookupReader(PosData<T>[] data) {
+				this.data = data;
+				current = new PosData<T>(-1, default);
+				nextIdx = 0;
+			}
+
+			public T Get(int pos) {
+				if (pos <= current.pos)
+					throw new ArgumentException($"Must read in ascending index order. Prev: {current.pos}, pos: {pos}");
+
+				while (nextIdx < data.Length && data[nextIdx].pos <= pos)
+					current = data[nextIdx++];
+
+				return current.value;
+			}
+		}
+
+
+		public readonly int pos;
+		public readonly T value;
+
+		public int X => pos / Main.maxTilesY;
+		public int Y => pos % Main.maxTilesY;
+
+		public PosData(int pos, T value) {
+			this.pos = pos;
+			this.value = value;
+		}
+
+		public PosData(int x, int y, T value) : this(PosData.CoordsToPos(x, y), value) { }
+	}
+
+	public static class PosData
+	{
+		public static int CoordsToPos(int x, int y) => x * Main.maxTilesY + y;
+
+		/// <summary>
+		/// Searches for the interval posMap[i].posID < posID < posMap[i + 1].posID and is preferable to NearbyBinarySearchPosMap for ordered data.
+		/// </summary>
+		/// <param name="posMap"></param>
+		/// <param name="posID"></param>
+		/// <returns></returns>
+		public static int FindIndex<T>(this PosData<T>[] posMap, int pos) {
+			int minimum = -1, maximum = posMap.Length;
+			while (maximum - minimum > 1) {
+				int split = (minimum + maximum) / 2;
+
+				if (posMap[split].pos <= pos)
```

Comment:
Reverse this, again per original review, its wrong for an interval search as the exit condition takes wasted cycles to reach